chore(vrl): add development design document (#8735)
vectordotdev/vector: lib/vrl/DESIGN.md

# Vector Remap Language: Guiding Design Principles

This document describes the high-level goals and directions of the Vector Remap
Language (VRL). It is intended to help guide discussions and decisions made
during the development of VRL itself. The document is less relevant to _users_
of VRL, the [language documentation][docs] exists for that audience.

[docs]: https://vrl.dev

## Table of Contents

* [The Zen of VRL](#the-zen-of-vrl)
* [Why VRL](#why-vrl)
* [Target Audience](#target-audience)
* [Language Limits](#language-limits)
* [Conventions](#conventions)
  * [Functions](#functions)
    * [Composition](#composition)
    * [Naming](#naming)
    * [Return Types](#return-types)
    * [Mutability](#mutability)
    * [Fallibility](#fallibility)
    * [Signatures](#signatures)
  * [Errors](#errors)
* [Patterns](#patterns)
  * [Error Chaining](#error-chaining)

> ## The Zen of VRL

Reviewer: To expand on this. These two sections should align with this doc. There's probably some thing in here that should be copied over:

* [VRL Features](https://vector.dev/docs/reference/vrl/#features)
* [VRL Principles](https://vector.dev/docs/reference/vrl/#principles)

> ## The Zen of VRL

Author: I added your two paragraphs, and removed the bullet list. I'll follow up with whatever else we're missing from those two links in separate PRs.

In the spirit of [The Zen of Python][PEP20].

- Beautiful is better than ugly.
- Explicit is better than implicit.
- Simple is better than complex.
- Sparse is better than dense.
- Readability trumps writing convenience.
- Special cases aren't special enough to break the rules.
- Although practicality beats purity.
- Function calls over syntax.
- Errors may never occur at run-time.
- Unless explicitly marked as such.
- There should be one — and preferably only one — obvious way to do it.
- Performance matters.
- Provide observability remapping solutions, nothing else.

[PEP20]: https://www.python.org/dev/peps/pep-0020/

## Why VRL

VRL exists to solve the need for a _simple_ and _performant_ **domain specific
language** (DSL) to remap observability data.

> Its purpose is to hit a sweet spot between Turing complete languages such as
> Lua, and static transforms such as Vector's `rename_fields`. It should be
> flexible enough to cover most remapping use-cases, without requiring operators
> to write complex scripts that are difficult to reason about and incur
> a significant performance penalty.

Reviewer: ```suggestion
Its purpose is to hit a sweet spot between Turing complete languages such as
Lua, and static transforms such as Vector's old `rename_fields`. It should be
flexible enough to cover most remapping use-cases, without requiring the
flexibility and downsides of a full programming language, such as poor readability
and performance.
```

I think these changes a little more specific with the goals. Take it or leave it.

See the [introduction blog post][blog] for more details on the **why**.

[blog]: https://vector.dev/blog/vector-remap-language/#preamble

> ## Target Audience

Reviewer: Excellent section

VRL has a specialized purpose of solving observability data remapping needs.
Because of this purpose, the language is mostly used by people who manage
infrastructure at their organisations. The role of this group is usually
referred to as "operator", "devops" (Development and Operations) or "SRE" (Site
Reliability Engineer).

One common generalization of this group is that they are focused on maintaining
infrastructure within an organization, and often write and maintain their own
software to achieve this goal. They are adept enough at programming to achieve
their goals, but have no need or desire to be as skilled in programming as
dedicated software engineers, because their time is best spent elsewhere.

As with everything, there are exceptions to the rule, and many people in this
group _are_ highly skilled software engineers, but VRL must capture the largest
possible segment of this group, and should therefor be limited in complexity.

> Therefor, when extending the feature set of VRL, design **the feature for the

Reviewer 2: ```suggestion
Therefore, when extending the feature set of VRL, design **the feature for the
```

Author: Thanks for the heads-up.

ref: https://grammarist.com/usage/therefore-vs-therefor/

Reviewer 2: TIL therefor

intended target audience**, which will likely mean choosing different trade-offs
than you'd make if you were to design the feature for your personal needs.

## Language Limits

There are a number of features that we've so far rejected to implement in the
language. This might change in the future, but there should be a good reason to
do so.

- modules (see: [#5507][]
- classes
- user-defined functions
- `goto` statements

> - network calls (see [#4517][])

Reviewer 2: Notably I think we may be loosening this a bit if we introduce [`reverse_dns`](#4517) via https://github.com/timberio/vector/pull/8717

> - network calls (see [#4517][])

Reviewer: Can we provide short justifications for each? They should connect back to our principles.

[#5507]: https://github.com/timberio/vector/issues/5507
[#4517]: https://github.com/timberio/vector/issues/4517#issuecomment-754160338

## Conventions

### Functions

#### Composition

- Favor function composition over single-purpose functions.

- If a problem needs to be solved in multiple steps, consider adding
  single-purpose functions for each individual step.

- If useability or readability is hurt by composition, favor single-purpose
  functions.

#### Naming

- Function names are lower-cased (e.g. `round`).

- Multi-name functions use underscore (`_`) as a separator (e.g.
  `encode_key_value`).

- Favor explicit verbose names over terse ones (e.g.
  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).

- Use `parse_*` for string decoding functions (e.g. `parse_json` and
  `parse_grok`).

> - Use `encode_*` for string encoding functions (e.g. `encode_base64`).

Author: It should be noted that some decoding functions are named `decode_*` (`decode_base64` and `decode_percent`).

Ideally, we'd rectify this, but that's a breaking change. We can create issues for when a current implementation goes against the design principles so that we can do a sweep of breaking changes before we hit 1.0.

Alternatively, we accept this, document it, and keep things as is, even when we release 1.0.

Reviewer 2: I think there is some distinction here between functions that parse a string into some other type (like `parse_json` or `parse_int`) and functions that turn a string into another string (`decode_base64` or `decode_percent`).

To me `parse_base64` reads pretty weird given no "parsing" is really happening, just translation. I think `parse` makes sense for only for structured data.

It does makes `encode` a little confusing since it is used as the opposite of both `decode` and `parse` when you might only expect it to be the opposite of `decode`. I might suggest we use `to_X` like `to_json` and `to_key_value` but this would overload `to_X` given it is also used for converting to primitives. We could go the Go/Java route of `unmarshal` and `marshal`?

> Ideally, we'd rectify this, but that's a breaking change. We can create issues for when a current implementation goes against the design principles so that we can do a sweep of breaking changes before we hit 1.0.

I think we could also just alias the functions to rename them to the desired scheme; dropping the deprecated aliases at 1.0.

Reviewer 3: 💭 `marshal`/`unmarshal` seem natural given my history in Go - but it is a larger change (aliasing seems reasonable)

- Use `to_*` to convert from one type to another (e.g. `to_bool`).

- Use `format_*` for string formatting functions (e.g. `format_timestamp` and

>   `format_number`).

Reviewer: I think we can generalize these and say that functions should be preceded with their function category for organization and discovery. You could explicitly list each, but that feels cumbersome and unnecessary.

#### Return Types

> - Return boolean from `is_*` functions (e.g. `is_string`).

Reviewer: You call out `is_*` functions here, but don't mention anything about `is_*` functions in the naming section.

> - Return an error when `parse_*` functions fail to decode the string.

Author: All but one `parse_*` functions are fallible. `parse_query_string` being the exception.

> - Return an error when `parse_*` functions fail to decode the string.

Reviewer: This seems weirdly specific. Don't other functions types return errors? Is it better to generalize this qualification?

#### Mutability

As a general rule, functions never mutate values.

Favor this:

```coffee
# explicitly assign the parsed JSON to `.message`
.message = parse_json(.message)
```

Over this:

```coffee
# mutate `.message` in place
parse_json(.message)
```

There are exceptions to this rule (such as the `del` function), but they are
limited, and additional exceptions should be well reasoned.

#### Fallibility

- A function must be marked as fallible if it can fail in any way.

- Design a function with the goal of making it infallible.

- But don't hide fallibility for the sake of persuing the previous rule.

- A function implementation can assume it receives the argument type it has
  defined.

- `parse_*` functions should almost always error when used incorrectly.

> - `to_*` functions must never fail.

Author: This isn't currently true because we decided to have `to_*` functions also take a string, meaning it does _both_ decoding and type conversion.

We can either update this rule to match reality, or (at some point before 1.0) do a breaking change to update the functions to match this rule.

Reviewer 3: 💭  would that end up being something like:

```vrl
if is_integer(.field) {
    .field = int(.field)
}
```

#### Signatures

- Functions can have zero or more parameters (e.g. `now()` and `assert(true)`).

- Argument naming follows the same conventions as function naming.

- For one or more parameters, the first parameter must be the "target" of the
  function (e.g. `parse_regex(target: <string>, pattern: <regex>)`).

> - The first parameter must therefor almost always be named `target`.

Author: Again, not true currently, we (almost always) name the first parameter `value`. When writing this down, it felt like `target` was the better name for it, but that would require updating all functions.

Same as previous comments, we can track this and fix it before 1.0, or keep as is and update this guide to rename `target` to `value`.

Reviewer 2: `target` says to me that the value _will_ be mutated. My personal vote is to keep it `value`.

It looks like the exceptions to using `value` are currently:

- `assert(condition: <boolean>, message: <string>)`
- `assert_eq(left: <any>, right: <any>, message: <string>)
- `del(path: <path>)`
- `exists(path: <path>)`
- `ip_cidr_contains(cidr: <string>, ip: <string>)`
- `ip_subnet(ip: <string>, subnet: <string>)`
- `ip_to_ipv6(ip: <string>)`
- `ipv6_to_ipv4(ip: <string>)`
- `merge(to: <string>, from: <object>, deep: <boolean>)`
- `unnest(path: <path>)`
- `get_env_var(name: <string>)`

Given those examples, I think we could amend this rule slightly to be:

- Call the first parameter `path` if it can only be a VRL path. I think this makes it clearer when paths are required vs. values.
- Otherwise, `value` unless you have a good reason not to (like `merge`, `assert`, and `get_env_var`)

I think we could update the `ip` functions to use `value` as an alias.

Reviewer 3: I'm also pro `value` - though I'd argue sometimes `path` and `value` overlap some...

Author: Yeah, you both make a good argument, I'll make the change.

> - The first parameter must therefor almost always be named `target`.

Reviewer 2: ```suggestion
- The first parameter must therefore almost always be named `target`.
```

- TODO

### Errors

TODO

## Patterns

The following is a list of patterns we've experienced while writing and using
VRL. This section of the document is intended to be updated frequently with new
insights.

These insights are meant to guide future design decisions, and shape our
thinking of the language as it matures and we learn more about its strengths and
weaknesses from our users.

### Error Chaining

Observability data can be structured in unexpected ways. A common pattern is to
try to decode the data in one way, only to try a different decoder if the first
one failed.

This pattern uses [function calls][] and [error coalescing][] to achieve its
goal:

```coffee
data = parse_json(.message) ??
       parse_nginx_log(.message) ??
       parse_apache_log(.message) ??
       { "error": "invalid data format" }
```

> TODO

Author: If you've seen any common patterns in the wild that are worth mentioning here, please let me know!

Reviewer 2: 👍 I like this pattern.

I am wondering if we'll want to introduce "pipelining" at some point like:

```
data = decode_base64(.message) |> parse_json
```

That is really just syntax sugar though.

Author: Pipe operator support is tracked in https://github.com/timberio/vector/issues/5431, but I don't think there's anything to note here, since it isn't actually part of the language yet.

intended target audience**, which will likely mean choosing different trade-offs
than you'd make if you were to design the feature for your personal needs.

## Language Limits

There are a number of features that we've so far rejected to implement in the
language. This might change in the future, but there should be a good reason to
do so.

- modules (see: [#5507][])

  So far, we've had no need to split up functions over multiple modules. The
  function naming rules make it so that most functions are already grouped by
  their logical usage patterns.

- classes

  Given that VRL is a simple DSL, and that any indirection in a program's source
  can lead to confusion, we've decided against introducing the concept of
  classes, and instead focused on the usage of function calls to solve operator
  needs.

- user-defined functions

  User-defined functions again produce indirection. While it _might_ be useful
  to some extremely large use-cases, in most cases, allowing people to read
  a program from top to bottom without having to jump around is more clear in
  the context within which VRL is used (remapping).

> - network calls (see [#4517][] and [#8717][])

Reviewer: Could this be expanded to general I/O, or do you think that differs?

Reviewer 2: 🤔 I think as we add enrichment functionality to VRL this will change. Imagining that we add an enrichment table for Redis and then call out to it from VRL, this could involve a network call. Granted VRL won't be doing the call, the enrichment table will, but it still affects throughput in the same way as VRL doing the network call itself.

  In order to avoid performance footguns, we want to ensure each function is as
  performant as it can be, and there's no way to use functions in such a way
  that performance of a VRL program tanks. We might introduce network calls at
  some point, if we find a good caching solution to solve most of our concerns,
  but so far we've avoided any network calls inside our stdlib.

[#5507]: https://github.com/timberio/vector/issues/5507
[#4517]: https://github.com/timberio/vector/issues/4517#issuecomment-754160338
[#8717]: https://github.com/timberio/vector/pull/8717

## Conventions

### Syntax

Keep VRL as syntax-light as possible. The less symbols, the more readable a VRL
program is.

Use functions whenever possible, and only introduce new syntax if a common
pattern warrants a more convenient syntax-based solution, or functions are too
limited in their capability to solve the required need.

### Functions

#### Composition

- Favor function composition over single-purpose functions.

- If a problem needs to be solved in multiple steps, consider adding

>   single-purpose functions for each individual step.

Author: That is a fair point. I think it's a balancing act between keeping a program understandable for the target audience, and trying to squeeze as much performance out of it as possible.

I also think that — as long as we see a path forward solving this within the compiler itself (e.g. some kind of optimization pass), then we should avoid burdening operators with having to search through hundreds of single-purpose functions to fit their needs.

I'll add a chapter dedicated to performance, and the trade-offs we need to make with each decision.

- If useability or readability is hurt by composition, favor single-purpose
  functions.

#### Naming

- Function names are lower-cased (e.g. `round`).

- Multi-name functions use underscore (`_`) as a separator (e.g.
  `encode_key_value`).

- Favor explicit verbose names over terse ones (e.g.
  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).

- Functions should be preceded with their function category for organization and
  discovery.

  - Use `parse_*` for string to type decoding functions (e.g. `parse_json` and
    `parse_grok`).
  
  - Use `decode_*` for string to string decoding functions (e.g.
    `decode_base64`).
  
  - Use `encode_*` for string encoding functions (e.g. `encode_base64`).
  
  - Use `to_*` to convert from one type to another (e.g. `to_bool`).

  - Use `is_*` to determine if the provided value is of a given type (e.g.
    `is_string` or `is_json`).
  
  - Use `format_*` for string formatting functions (e.g. `format_timestamp` and
    `format_number`).

> 

Reviewer 4: We should add:

```markdown
- Use `get_*` for functions that return a single result, or error if a single result is not found.
- Use `find_*` when multiple possible results are returned in an array.
```

#### Return Types

- Return boolean from `is_*` functions (e.g. `is_string`).

- Return a string from `encode_*` functions (e.g. `encode_base64`).

- Return an error whenever the function can fail at runtime.

#### Mutability

As a general rule, functions never mutate values.

Favor this:

```coffee
# explicitly assign the parsed JSON to `.message`

> .message = parse_json(.message)

Reviewer 3: Do we want to keep the examples valid VRL?

Author: It's a snippet of a valid VRL program, but I don't think we need to show the entire program to explain a particular pattern, especially given that this is focussed on developers, not users.

```

Over this:

```coffee
# mutate `.message` in place
parse_json(.message)
```

There are exceptions to this rule (such as the `del` function), but they are
limited, and additional exceptions should be well reasoned.

#### Fallibility

- A function must be marked as fallible if it can fail in any way.

- Design a function with the goal of making it infallible.

- But don't hide fallibility for the sake of persuing the previous rule.

- A function implementation can assume it receives the argument type it has
  defined.

- `parse_*` functions should almost always error when used incorrectly.

> 

Reviewer 4: ```markdown
- `get_*` functions should fail when it can't find a single result to return.
```

- `to_*` functions must never fail.

#### Signatures

- Functions can have zero or more parameters (e.g. `now()` and `assert(true)`).

- Argument naming follows the same conventions as function naming.

- For one or more parameters, the first parameter must be the "value" the
  function is acting on (e.g. `parse_regex(value: <string>, pattern: <regex>)`).

- The first parameter must therefore almost always be named `value`.

- The exception to this is when you're dealing with an actual VRL path (e.g.

>   `del(path)`) or in special cases such as `assert`.

Reviewer 3: This strikes me as it seems generally usage is `parse_regex(.message, ...)` where the `value` is extracted from the `path`

Author: Whether a function extracts the value from a path depends on the function implementation.

Almost always, yes, a function will resolve the path to a value, but in a few cases where we want mutability (such as `del`), the function doesn't actually resolve the path to a value, but instead asks the VRL runtime to _delete the value at the given path_.

Reviewer 3: 👍 just something that came to mind as I was pondering the distinction between our "first parameters"

Author: Yep, I agree it can be confusing at first, especially as a developer. For users, it should (hopefully) result in a "it just works" situation.

### Errors

> - All errors are caught at compile-time by the compiler.

Reviewer 5: It might be worth clarifying what is covered by "all errors" here.

# Vector Remap Language: Guiding Design Principles

This document describes the high-level goals and directions of the Vector Remap
Language (VRL). It is intended to help guide discussions and decisions made
during the development of VRL itself. The document is less relevant to _users_
of VRL, the [language documentation][docs] exists for that audience.

[docs]: https://vrl.dev

## Table of Contents

> <!-- vim-markdown-toc GFM -->

Reviewer 4: Totally off topic, but what does this line do?

Author: They are markers for the Vim plugin I use to generate a table of contents. I usually keep it in while I develop the RFC/document so that the TOC gets auto-updated as I edit the rest of the document, and then remove them before I merge the document.

(I just removed them in the final commit, as I'm getting ready to merge)

do so.

- modules (see: [#5507][])

  So far, we've had no need to split up functions over multiple modules. The
  function naming rules make it so that most functions are already grouped by
  their logical usage patterns.

- classes

  Given that VRL is a simple DSL, and that any indirection in a program's source
  can lead to confusion, we've decided against introducing the concept of
  classes, and instead focused on the usage of function calls to solve operator
  needs.

- user-defined functions

  User-defined functions again produce indirection. While it _might_ be useful
  to some extremely large use-cases, in most cases, allowing people to read
  a program from top to bottom without having to jump around is more clear in
  the context within which VRL is used.

- network calls (see [#4517][] and [#8717][])

  In order to avoid performance foot guns, we want to ensure each function is as
  performant as it can be, and there's no way to use functions in such a way
  that performance of a VRL program tanks. We might introduce network calls at
  some point, if we find a good caching solution to solve most of our concerns,

>   but so far we've avoided any network calls inside our stdlib.

Reviewer 4: I think the biggest issue here is async. If we want to start introducing IO we will need to rewire VRL calls to be async. If we do it well it shouldn't affect performance too much since the runtime can be handling other events while waiting on IO.

Author: That's a fair point. I don't think it fits this document, but worth keeping in mind.

  fail.

- A function _should not_ be marked as fallible if it receives the wrong
  argument type. This is handled by the compiler.

- Errors should contain explicit messages detailing what went wrong, and how the
  operator can solve the problem.

### Functions

#### Composition

- Favor function composition over single-purpose functions.

- If a problem needs to be solved in multiple steps, consider adding
  single-purpose functions for each individual step.

- If usability or readability is hurt by composition, favor single-purpose
  functions.

#### Naming

- Function names are lower-cased (e.g. `round`).

- Multi-name functions use underscore (`_`) as a separator (e.g.
  `encode_key_value`).

- Favor explicit verbose names over terse ones (e.g.
  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).

- Functions should be preceded with their function category for organization and
  discovery.

  - Use `parse_*` for functions that decode a string to another type of data
    (e.g. `parse_json` and `parse_grok`).
  
  - Use `decode_*` for string to string decoding functions (e.g.
    `decode_base64`).
  

>   - Use `encode_*` for string encoding functions (e.g. `encode_base64`).

Reviewer 2: I'm still partial to the `unmarshal` / `marshal` terminology for converting from/to serialized structured data (like `unmarshal_json`); leaving `encode` / `decode` for converting bytes to each other, but I'm ok if we want to leave that discussion for later. What you have here does codify the existing patterns.

Author: Created https://github.com/vectordotdev/vector/issues/9405 to track this.
